src/qttools/datastructures/dsbsparse.py

```diff
@@ -1,244 +1,538 @@
+        """Matrix multiplication of two DSBSparse matrices."""
+        ...
 
-    def diagonal(self) -> np.ndarray:
-        """Returns the diagonal of the matrix."""
-        return np.hstack(
-            [
-                np.diagonal(self[b, b], axis1=-2, axis2=-1)
-                for b in range(self.num_blocks)
-            ]
-        )
+    def block_diagonal(self, offset: int = 0) -> list[ArrayLike]:
+        """Returns the block diagonal of the matrix.
+
+        Parameters
+        ----------
+        offset : int, optional
+            Offset from the main diagonal. Positive values indicate
+            superdiagonals, negative values indicate subdiagonals.
+            Default is 0.
+
+        Returns
+        -------
+        blocks : list
+            List of block diagonal elements. The length of the list is
+            the number of blocks on the main diagonal minus the offset.
+            Depending on return_dense, the elements are either sparse
+            or dense arrays.
+
+        """
+        blocks = []
+        for b in range(self.num_blocks - abs(offset)):
+            blocks.append(self.blocks[b, b + offset])
+
+        return blocks
+
+    def diagonal(self) -> ArrayLike:
+        """Returns the diagonal elements of the matrix.
+
+        This temporarily sets the return_dense state to True.
+
+        Returns
+        -------
+        diagonal : array_like
+            The diagonal elements of the matrix.
+
+        """
+        # Store the current return_dense state and set it to True.
+        original_return_dense = self.return_dense
+        self.return_dense = True
+
+        diagonals = []
+        for b in range(self.num_blocks):
+            diagonals.append(xp.diagonal(self.blocks[b, b], axis1=-2, axis2=-1))
+
+        # Restore the original return_dense state.
+        self.return_dense = original_return_dense
+        return xp.hstack(diagonals)
+
+    def _dtranspose(self, block_axis: int, concatenate_axis: int) -> None:
+        """Performs the distributed transposition of the data.
+
+        This is a helper method that performs the distributed transposition
+        depending on the current distribution state.
+
+        Parameters
+        ----------
+        block_axis : int
+            The axis along which the blocks view is created.
+        concatenate_axis : int
+            The axis along which the received blocks are concatenated.
+
+        """
+        # old_shape = self._data.shape
+        # new_shape = (
+        #     old_shape[0] // comm.size,
+        #     *old_shape[1:-1],
+        #     old_shape[-1] * comm.size,
+        # )
+
+        self._data = _block_view(self._data, axis=block_axis)
+        # We need to make sure that the block-view is memory-contiguous.
+        # This does nothing if the data is already contiguous.
+        self._data = xp.ascontiguousarray(self._data)
+
+        synchronize_current_stream()
+        comm.Alltoall(MPI.IN_PLACE, self._data)
+
```

Comment:
@vincent-maillou for the `cupy` backend, should we change this by just transferring the data to host for now? We can try to do more complicated stuff, like detecting if MPI is GPU aware later on... 

```suggestion
        synchronize_current_stream()
        comm.Alltoall(MPI.IN_PLACE, get_host(self._data))
```